xapian-letor/api/letor_internal.cc

```diff
@@ -315,19 +313,125 @@ Letor::Internal::prepare_training_file(const string & queryfile, const string &
 	Xapian::MSet mset = enquire.get_mset(0, msetsize);
 
 	std::vector<FeatureVector> fvv_mset = flist.create_feature_vectors(mset, query, letor_db);
+	std::vector<FeatureVector> fvv_qrel;
 
-	// Set labels from qrel file to FeatureVectors
 	int k = 0;
 	for (Xapian::MSetIterator i = mset.begin(); i != mset.end(); ++i) {
 
 	    Xapian::Document doc = i.get_document();
-	    fvv_mset[k++].set_label(getlabel(doc, qid));
+	    int label = getlabel(doc, qid);
+	    if (label != -1) { // only add FeatureVector which is found in the qrel file
+		fvv_mset[k].set_label(label);
+		fvv_qrel.push_back(fvv_mset[k]);
+	    }
+	    k += 1;
 	}
 
-	write_to_file(fvv_mset, qid, train_file);
+	write_to_file(fvv_qrel, qid, train_file);
 
     }
     myfile1.close();
     train_file.close();
     cout << "Training file stored as: \"" << filename << "\"" << endl;
 }
+
+void
+Letor::Internal::letor_score(const std::string & query_file,
+			     const std::string & qrel_file,
+			     const std::string & model_file,
+			     Xapian::doccount msetsize,
+			     Xapian::FeatureList & flist) {
+
+    ranker->load_model_from_file(model_file.c_str());
+
+    Xapian::SimpleStopper mystopper(sw, sw + sizeof(sw) / sizeof(sw[0]));
+    Xapian::Stem stemmer("english");
+
+    Xapian::QueryParser parser;
+    parser.add_prefix("title", "S");
+    parser.add_prefix("subject", "S");
+
+    parser.set_database(letor_db);
+    parser.set_default_op(Xapian::Query::OP_OR);
+    parser.set_stemmer(stemmer);
+    parser.set_stemming_strategy(Xapian::QueryParser::STEM_SOME);
+    parser.set_stopper(&mystopper);
+
+    this->qrel = load_relevance(qrel_file);
+
+    string str1;
+    ifstream queryfile;
+    queryfile.open(query_file.c_str(), ios::in);
+
+    if(!queryfile.good()){
+	cout << "No Query file found"<<endl;
+	throw FileNotFound();
+    }
+
+    double total_score = 0;
+    int num_queries = 0;
+
+    while (!queryfile.eof()) {           //reading all the queries line by line from the query file
+
+	getline(queryfile, str1);
+
+	if (str1.empty()) {
+	    break;
+	}
+
+	string qid = str1.substr(0, (int)str1.find(" "));
+	string querystr = str1.substr((int)str1.find("'")+1, (str1.length() - ((int)str1.find("'") + 2)));
```

Comment:
Spaces missing around the first `+`.

The parsing here is rather crude and lacking in error checking (there's nothing to require the `'` to be after the space for example, and it'll happily continue even if there's no '`' in the line).  It also searches for the same `'` twice, though perhaps the compiler can spot that it doesn't need to repeat that search.  I guess these can get dealt with later though.